components/layout_2020/taffy/construct.rs

```diff
@@ -0,0 +1,240 @@
+use crate::style_ext::DisplayGeneratingBox;
+
+impl TaffyContainer {
+    pub fn construct<'dom>(
+        context: &LayoutContext,
+        info: &NodeAndStyleInfo<impl NodeExt<'dom>>,
+        contents: NonReplacedContents,
+        propagated_text_decoration_line: TextDecorationLine,
+    ) -> Self {
+        let text_decoration_line =
+            propagated_text_decoration_line | info.style.clone_text_decoration_line();
+        let mut builder = TaffyContainerBuilder {
+            context,
+            info,
+            text_decoration_line,
+            contiguous_text_runs: Vec::new(),
+            jobs: Vec::new(),
+            has_text_runs: false,
+        };
+        contents.traverse(context, info, &mut builder);
+        builder.finish()
+    }
+}
+
+/// <https://drafts.csswg.org/css-flexbox/#flex-items>
+struct TaffyContainerBuilder<'a, 'dom, Node> {
+    context: &'a LayoutContext<'a>,
+    info: &'a NodeAndStyleInfo<Node>,
+    text_decoration_line: TextDecorationLine,
+    contiguous_text_runs: Vec<FlexTextRun<'dom, Node>>,
+    /// To be run in parallel with rayon in `finish`
+    jobs: Vec<FlexLevelJob<'dom, Node>>,
+    has_text_runs: bool,
+}
+
+enum FlexLevelJob<'dom, Node> {
+    /// Or pseudo-element
+    Element {
+        info: NodeAndStyleInfo<Node>,
+        display: DisplayGeneratingBox,
+        contents: Contents,
+        box_slot: BoxSlot<'dom>,
+    },
+    TextRuns(Vec<FlexTextRun<'dom, Node>>),
+}
+
+struct FlexTextRun<'dom, Node> {
+    info: NodeAndStyleInfo<Node>,
+    text: Cow<'dom, str>,
+}
+
+impl<'a, 'dom, Node: 'dom> TraversalHandler<'dom, Node> for TaffyContainerBuilder<'a, 'dom, Node>
+where
+    Node: NodeExt<'dom>,
+{
+    fn handle_text(&mut self, info: &NodeAndStyleInfo<Node>, text: Cow<'dom, str>) {
+        self.contiguous_text_runs.push(FlexTextRun {
+            info: info.clone(),
+            text,
+        })
+    }
+
+    /// Or pseudo-element
+    fn handle_element(
+        &mut self,
+        info: &NodeAndStyleInfo<Node>,
+        display: DisplayGeneratingBox,
+        contents: Contents,
+        box_slot: BoxSlot<'dom>,
+    ) {
+        // FIXME: are text runs considered "contiguous" if they are only separated
+        // by an out-of-flow abspos element?
+        // (That is, are they wrapped in the same anonymous flex item, or each its own?)
+        self.wrap_any_text_in_anonymous_block_container();
+
+        self.jobs.push(FlexLevelJob::Element {
+            info: info.clone(),
+            display,
+            contents,
+            box_slot,
+        })
+    }
+}
+
+/// <https://drafts.csswg.org/css-text/#white-space>
+fn is_only_document_white_space<Node>(run: &FlexTextRun<'_, Node>) -> bool {
+    // FIXME: is this the right definition? See
+    // https://github.com/w3c/csswg-drafts/issues/5146
+    // https://github.com/w3c/csswg-drafts/issues/5147
+    run.text
+        .bytes()
+        .all(|byte| matches!(byte, b' ' | b'\n' | b'\t'))
+}
```

Comment:
I would make the version of this from the flex code public and reuse it rather than copy-and-pasting here.